### PR TITLE
Moving the latest changes from Stash

### PR DESCRIPTION
**Pull request by danielmkn**
Moving the latest Transformer Lambda changes to GitHub repo. SAM template modifications. 
The deployment was successfully tested end to end.

### CODE REVIEW

File: transformer/app.js [hunk 1]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
     const accountId = process.env.USE_DEV_ACCOUNT_ID === 'true' ? process.env.DEV_ACCOUNT_ID : context.invokedFunctionArn.split(':')[4];
-    const parsedBody = JSON.parse(event.Records[0].body);
-    let findings = [];
-    const type = parsedBody.type.toLowerCase();
-    findings = transformIssue(parsedBody, type, accountId);
-    console.log(`Findings to send to Hub: ${JSON.stringify(findings)}`);
-    const hubResponse = await hubClient.send(new BatchImportFindingsCommand({ Findings: findings }));
-    console.log(`Security Hub response: ${JSON.stringify(hubResponse)}`);
+    const findingsCollection = event.Records.map((sqsEvent) => {
+      const parsedBody = JSON.parse(sqsEvent.body);
```
Comment (alexhung) [inline transformer/app.js:202]: Let's rename `parseBody` to `issue`.

File: transformer/app.js [hunk 2]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+    });
+    console.debug(`Complete findings list: ${JSON.stringify(findingsCollection.flat())}`);
+    const completeIdsCollection = findingsCollection.flat().map((finding) => finding.Id);
+    console.debug(`Unfiltered finding IDs collection: ${JSON.stringify(completeIdsCollection)}`);
+    const existingFindingIDsInDB = await verifyFindingIds(completeIdsCollection);
+    console.debug(`Existing finding IDs in the DB: ${JSON.stringify(existingFindingIDsInDB)}`);
+    const newFindingIDs = completeIdsCollection.filter((id) => !existingFindingIDsInDB.includes(id));
+    console.debug(`New findings to import (IDs only): ${JSON.stringify(newFindingIDs)}`);
+    const existingFindingsToUpdate = findingsCollection.flat().filter((item) => existingFindingIDsInDB.includes(item.Id));
+    const newFindingsToImport = findingsCollection.flat().filter((item) => !existingFindingIDsInDB.includes(item.Id));
```
Comment (alexhung) [inline transformer/app.js:213]: Please add blank lines between these groups of code to make them easier to read.

Comment (alexhung) [reply transformer/app.js:213]: This group of code is good candidate to be extracted into a function. Between L205 to L213, we really only care about the 2 list of findings. The code in between are details that can be hidden inside a function.

File: transformer/app.js [hunk 3]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+    const timestamp = new Date().getTime().toString();
+    try {
+      promises.push(ddbClient.send(new PutCommand(writeParams(id, timestamp))));
+      console.log(`Writing ID to DB: ${finding.Id}`);
+    } catch (err) {
+      console.error(err);
+    }
+  });
+  const results = await Promise.allSettled(promises);
+  const resultsFulfilled = results.filter((result) => (result.status === 'fulfilled')).map((result) => result.status);
```
Comment (alexhung) [inline transformer/app.js:152]: Shouldn't the return value in the map function be `result.value`?

File: transformer/app.js [hunk 4]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+  const resultsFailed = results.filter((result) => (result.status !== 'fulfilled')).map((result) => result.message);
```
Comment (alexhung) [inline transformer/app.js:153]: The return value in the map function should be `result.reason`, right?

Comment (alexhung) [inline transformer/app.js:153]: Let's use `result.status === 'rejected'` instead to be clear.

File: transformer/app.js [hunk 5]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+    return result.value.Items[0].ID.S;
+  });
+};
+
+const sendUpdateCommand = async (updateBody) => {
+  const promises = updateBody.map((body) => hubClient.send(new BatchUpdateFindingsCommand(body)));
+  const results = await Promise.allSettled(promises);
+  console.debug(`Unfiltered results (promises): ${JSON.stringify(results)}`);
+  const resultsFulfilled = results.filter((result) => (result.status === 'fulfilled')).map((result) => result.value.ProcessedFindings);
+  const resultsFailed = results.filter((result) => (result.status !== 'fulfilled')).map((result) => result.value.ProcessedFindings);
```
Comment (alexhung) [inline transformer/app.js:186]: Let's use `result.status === 'rejected'` instead to be clear.

File: transformer/app.js (hunk 2 above)
Comment (alexhung) [inline transformer/app.js:213]: `.flat()` is called multiple times. We should only call it once at L204.

**Review by alexhung — CHANGES_REQUESTED** [review]
See comments. May have more feedback once this batch of changes are completed.

File: eventProcessor/app.js [hunk 6]
```diff
@@ -29,7 +29,8 @@ const asyncLambdaInvoke = async (issuesChunks) => {
 
 export async function lambdaHandler(event) {
   const xrayEvent = JSON.parse(event.body);
-  const hostName = event.headers['X-Forwarded-For'];
+  console.debug(`Event body: ${event.body}`);
+  const hostName = event.headers.Hostname; // must be set as a custom header in Xray Webhook
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply eventProcessor/app.js:33]: Added. Fallback value is `hostname-was-not-set`.

File: transformer/app.js [hunk 7]
```diff
@@ -1,6 +1,22 @@
...
 
+const ddbClient = new DynamoDBClient();
 const hubClient = new SecurityHubClient();
+const marshallOptions = {
+  convertEmptyValues: false, // false, by default.
+  removeUndefinedValues: false, // false, by default.
+  convertClassInstanceToMap: false, // false, by default.
+};
+
+const unmarshallOptions = {
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:13]: Inlined.

File: transformer/app.js [hunk 8]
```diff
@@ -10,9 +26,9 @@ const SEVERITY_LABEL_LOOKUP = {
 };
 
 const TYPES_LOOKUP = {
-  security: 'Software and Configuration Checks/Vulnerabilities/CVE',
-  license: 'Software and Configuration Checks/Licenses/Compliance',
-  'operational risk': 'Software and Configuration Checks/Operational Risk',
+  security: ['Software and Configuration Checks/Vulnerabilities/CVE'],
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:29]: Array moved to the upstream function `getSeverityAndTypes`.

File: transformer/app.js [hunk 9]
```diff
@@ -64,7 +80,7 @@ const getCommonFields = (body, type, accountId) => ({
   ProductArn: `arn:aws:securityhub:${region}:${accountId}:product/${accountId}/default`,
   SchemaVersion: '2018-10-08',
   SourceUrl: `https://${body.host_name}/ui/watchesNew/edit/${body.watch_name}?activeTab=violations`,
-  Title: body.summary,
+  Title: `${body.summary.length > 256 ? `${(body.summary).substring(0, 251)}...` : body.summary}`,
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:83]: I'd rather keep it like this, I don't use Lodash in that lambda. No need to add an extra dependency.

File: transformer/app.js [hunk 10]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+    ID: id,
+    TIMESTAMP: timestamp,
+  },
+});
+
+const writeFindingsToDB = async (findingsCollection) => {
+  const promises = [];
+  findingsCollection.map(async (finding) => {
+    const id = finding.Id;
+    const timestamp = new Date().getTime().toString();
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:143]: Ok, moved.

File: transformer/app.js [hunk 11]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+      console.error(err);
+    }
+  });
+  const results = await Promise.allSettled(promises);
+  const resultsFulfilled = results.filter((result) => (result.status === 'fulfilled')).map((result) => result.status);
+  const resultsFailed = results.filter((result) => (result.status !== 'fulfilled')).map((result) => result.message);
+  return {
+    SuccessCount: resultsFulfilled.length,
+    SuccessfulRecords: resultsFulfilled,
+    FailedRecords: resultsFailed.length,
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:157]: Fixed, the idea was to do it in this order: 

```
    SuccessCount: resultsFulfilled.length,
    SuccessfulRecords: resultsFulfilled,
    FailedCount: resultsFailed.length,
    FailedRecords: resultsFailed,

File: transformer/app.js [hunk 12]
```diff
@@ -101,21 +118,122 @@ function transformIssue(body, type, accountId) {
...
+const writeFindingsToDB = async (findingsCollection) => {
+  const promises = [];
+  findingsCollection.map(async (finding) => {
+    const id = finding.Id;
+    const timestamp = new Date().getTime().toString();
+    try {
+      promises.push(ddbClient.send(new PutCommand(writeParams(id, timestamp))));
+      console.log(`Writing ID to DB: ${finding.Id}`);
+    } catch (err) {
+      console.error(err);
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:148]: Added.

File: transformer/app.js [hunk 13]
```diff
@@ -85,9 +101,10 @@ const getFindingProviderFields = (body, type) => ({
   FindingProviderFields: getSeverityAndTypes(body, type),
 });
 
-const getIdPrefix = (body, type) => (type === 'security' ? body.cve : body.summary);
+const getIdPrefix = (body, type) => (type === 'security' && body.cve ? body.cve : body.summary);
 
-function transformIssue(body, type, accountId) {
+function transformIssue(body, accountId) {
```
[the remark this answers is not in the document]
Comment (danielmkn) [reply transformer/app.js:106]: Ok

File: transformer/app.js (hunk 1 above)
Comment (danielmkn) [reply transformer/app.js:202]: Ok

File: transformer/app.js (hunk 3 above)
Comment (danielmkn) [reply transformer/app.js:152]: No, the resulting body of writing to the DB function doesn't have `value`, only `status`.

File: transformer/app.js (hunk 4 above)
Comment (danielmkn) [reply transformer/app.js:153]: It's `message` in this case. No `reason` in the promise body for this call.

Comment (danielmkn) [reply transformer/app.js:153]: Ok

File: transformer/app.js (hunk 5 above)
Comment (danielmkn) [reply transformer/app.js:186]: Ok

File: transformer/app.js (hunk 4 above)
Comment (alexhung) [reply transformer/app.js:153]: Are you 100% sure? The spec says `result` will be an object with `status` and `reason`/`value`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled#examples

File: transformer/app.js (hunk 2 above)
Comment (danielmkn) [reply transformer/app.js:213]: I think it's better to leave them as is, just add spaces for better readability.

File: transformer/app.js [hunk 14]
```diff
@@ -4,18 +4,11 @@ import { DynamoDBClient, QueryCommand } from '@aws-sdk/client-dynamodb';
...
-const marshallOptions = {
-  convertEmptyValues: false, // false, by default.
-  removeUndefinedValues: false, // false, by default.
-  convertClassInstanceToMap: false, // false, by default.
-};
 
-const unmarshallOptions = {
-  wrapNumbers: false, // false, by default.
+const translateConfig = {
+  convertEmptyValues: false, removeUndefinedValues: false, convertClassInstanceToMap: false, wrapNumbers: false,
```
Comment (alexhung) [inline transformer/app.js:9]: Please make this multilines.

**Review by alexhung — DISMISSED** [review]
One non-blocking change.

File: transformer/app.js (hunk 14 above)
Comment (danielmkn) [reply transformer/app.js:9]: sure

**Comment by danielmkn** [comment]
All comments are fixed.

**Pull request closed without merging**